syntax_sugar/infix.py

```diff
@@ -90,6 +90,17 @@ def __str__(self):
 def to(start, end):
     return To(start, end)
 
+@infix
+def by(to_object, step):
+    if to_object.end >= to_object.start and step < 0:
```

Comment:
You checked for `step` at both `__next__()` and `by()`. I think it could be better if we have all the checks in a property setter of `step` in the `To` class.